File: bundles/gitlab-eclipse-plugin/src/com/gitlab/eclipse/preferences/SecretStringFieldEditor.java
package com.gitlab.eclipse.preferences;

import java.util.Optional;

import org.eclipse.equinox.security.storage.StorageException;
import org.eclipse.jface.preference.StringFieldEditor;
import org.eclipse.swt.widgets.Composite;

import com.gitlab.eclipse.storage.SecretStorage;

public class SecretStringFieldEditor extends StringFieldEditor {
	private SecretStorage secretStorage;

	public SecretStringFieldEditor(SecretStorage secretStorage, String secretName, String labelText, Composite parent) {
		super(secretName, labelText, parent);
		this.secretStorage = secretStorage;
	}

	@Override
	protected void doLoad() {
		var textField = getTextControl();
		if (textField != null) {
			var value = Optional.ofNullable(secret())
					.map(s -> s.replaceAll(".", "•"))
					.orElse("");
			textField.setText(value);
			oldValue = value;
		}
	}

	@Override
	protected void doLoadDefault() {
		var textField = getTextControl();
		if (textField != null) {
			textField.setText("");
		}
		valueChanged();
	}

	@Override
	protected void doStore() {
		var textField = getTextControl();
		var text = textField.getText();
		if (text.matches("^•+$")) {
			// Avoid overwriting existing secrets with placeholder values. 
			return;
		}

		try {
			secretStorage.putSecret(getPreferenceName(), text);
		} catch (StorageException e) {
			e.printStackTrace();
		};
		doLoad();
	}
	
	@Override
	protected boolean doCheckState() {
		var text = getTextControl().getText();
		if (text.isEmpty()) {
			setErrorMessage(getLabelText() + " must not be empty");
			return false;
		}

		if (!text.matches("^[a-zA-Z0-9-]+$") && !text.matches("^•+$")) {
			setErrorMessage(getLabelText() + " must match pattern ^[a-zA-Z0-9-]+$");
			return false;
		}
		
		return true;
	}

	private String secret() {
		return secretStorage.getSecret(getPreferenceName());
	}
}

File: src/main/kotlin/com/gitlab/eclipse/preferences/SecretStringFieldEditor.kt
package com.gitlab.eclipse.preferences

import com.gitlab.eclipse.storage.SecretStorage
import org.eclipse.jface.preference.StringFieldEditor
import org.eclipse.swt.widgets.Composite
import java.util.*

class SecretStringFieldEditor(
    private val secretStorage: SecretStorage,
    secretName: String?,
    labelText: String?,
    parent: Composite?
) :
    StringFieldEditor(secretName, labelText, parent) {
    override fun doLoad() {
        val textField = textControl
        if (textField != null) {
            val value = Optional.ofNullable(secret())
                .map { s: String -> s.replace(".".toRegex(), "•") }
                .orElse("")
            textField.text = value
            oldValue = value
        }
    }

    override fun doLoadDefault() {
        val textField = textControl
        if (textField != null) {
            textField.text = ""
        }
        valueChanged()
    }

    override fun doStore() {
        val textField = textControl
        val text = textField.text
        if (text.matches("^•+$".toRegex())) {
            // Avoid overwriting existing secrets with placeholder values.
            return
        }

        secretStorage.putSecret(preferenceName, text)

        doLoad()
    }

    override fun doCheckState(): Boolean {
        val text = textControl.text
        if (text.isEmpty()) {
            errorMessage = "$labelText must not be empty"
            return false
        }

        if (!text.matches("^[a-zA-Z0-9-]+$".toRegex()) && !text.matches("^•+$".toRegex())) {
            errorMessage = "$labelText must match pattern ^[a-zA-Z0-9-]+$"
            return false
        }

        return true
    }

    private fun secret(): String? {
        return secretStorage.getSecret(preferenceName)
    }
}
